libzmix/bulletproofs_amcl/src/r1cs/gadgets/set_non_membership.rs

```diff
@@ -0,0 +1,192 @@
+use super::helper_constraints::constrain_lc_with_scalar;
+use super::helper_constraints::non_zero::is_nonzero_gadget;
+use crate::errors::R1CSError;
+use crate::r1cs::linear_combination::AllocatedQuantity;
+use crate::r1cs::{ConstraintSystem, LinearCombination, Prover, R1CSProof, Variable, Verifier};
+use amcl_wrapper::field_elem::FieldElement;
+use amcl_wrapper::group_elem::GroupElement;
+use amcl_wrapper::group_elem_g1::{G1Vector, G1};
+use merlin::Transcript;
+use rand::{CryptoRng, RngCore};
+
+/// Constraints for checking set membership check
+/// Create a new set with values being difference between the set value at that index and the value being proved a member.
+/// Now ensure that the product of members of this new set is not 0
+/// eg. Original set is (a, b, c, d, e). It is to be proved that x is not a member of set.
+/// Create new set (a-x, b-x, c-x, d-x, e-x). Now ensure product (a-x).(b-x).(c-x).(d-x).(e-x) != 0
+pub fn set_non_membership_gadget<CS: ConstraintSystem>(
+    cs: &mut CS,
+    v: AllocatedQuantity,
+    diff_vars: Vec<AllocatedQuantity>,
+    diff_inv_vars: Vec<AllocatedQuantity>,
+    set: &[FieldElement],
+) -> Result<(), R1CSError> {
+    let set_length = set.len();
+
+    for i in 0..set_length {
+        // Since `diff_vars[i]` is `set[i] - v`, `diff_vars[i]` + `v` should be `set[i]`
+        constrain_lc_with_scalar::<CS>(cs, diff_vars[i].variable + v.variable, &set[i]);
+        // Ensure `set[i] - v` is non-zero
+        is_nonzero_gadget(cs, diff_vars[i], diff_inv_vars[i])?;
```

Comment:
This seems to be the only difference with `set_membership`. Can we abstract this?

Reply:
Thats not the only difference. There is a product in set membership.